### DISCUSSION TITLE
FullyConnectedTensorProduct Feature Discrepancy in JAX vs. Torch #44

### DISCUSSION THREAD

**Discussion by tai-dang11** [e3nn/e3nn-jax#45] [Q&A] [answered]
To further clarify my question regarding FullyConnectedTensorProduct: In the PyTorch version, the weight parameter can accept both a single tensor of shape (..., self.weight_numel) or a list of tensors with shapes weight_shape / (...) + weight_shape when internal_weights is set to False. In my experience with the JAX version, it seems to primarily work with the list of tensors format. Is there support, or is there a planned feature, for also using a single tensor as a weight in the JAX version, similar to the PyTorch version?

Thank you again for your assistance!

<img width="1322" alt="Screenshot 2023-10-01 at 13 41 40" src="https://github.com/e3nn/e3nn-jax/assets/78414238/7f876945-3125-4d2e-872e-fab5c7ae1081">

**Comment by mariogeiger** [comment, marked as answer]
Here is an example on how to do that in e3nn-jax
```
def model(x):
    y = e3nn.tensor_product(x, x)
    lin = e3nn.FunctionalLinear(y.irreps, "0e + 1o")
    w = e3nn.haiku.MultiLayerPerceptron(
        [64, 64, lin.num_weights], jax.nn.silu, output_activation=False
    )(y.filter(keep="0e").array)
    return lin(w, y)


model = hk.without_apply_rng(hk.transform(model))
x = e3nn.normal("2x0e + 3x1o + 3x1e", jax.random.PRNGKey(0))

params = model.init(jax.random.PRNGKey(1), x)
y = model.apply(params, x)
```

**Reply by tai-dang11** [reply]
Hello,

Thank you for your previous insights regarding the equivalence of FullyConnectedTensorProduct to tensor_product followed by a Linear layer. I've attempted to employ this approach using Torch and it seems to function effectively. However, I'm encountering some discrepancies when trying to replicate this in Jax, specifically regarding the Linear layer in Flax.

While the tensor_product operations seem consistent between both Torch and Jax, I'm observing different weights and total parameters in the Linear layer between the two libraries, as illustrated below.

<img width="929" alt="Screenshot 2023-10-02 at 23 39 46" src="https://github.com/e3nn/e3nn-jax/assets/78414238/a98e2f0c-4d58-4a9e-8cb7-2db89462e702">
<img width="908" alt="Screenshot 2023-10-02 at 23 35 53" src="https://github.com/e3nn/e3nn-jax/assets/78414238/c767249a-a9b4-4efd-b766-d32a84a16741">

I’ve provided input data and additional context [in this shared drive](https://drive.google.com/drive/folders/1us0Llo1mxBpx-uPQLwAr1KtCIL6Vr0eI?usp=sharing) for further reference.

I'd appreciate it if you could help me understand if there's something I might be overlooking in the Jax/Flax implementation, or if there are inherent differences between the two that might account for these discrepancies.

Thank you for your time and assistance!

**Reply by mariogeiger** [reply]
Be careful feeding weights to `e3nn.Linear` has a different effect than using `e3nn.FunctionalLinear`. 

Look again carefully at the example I gave.

**Reply by tai-dang11** [reply]
Thank you so much. I got it now
